trax/models/research/funnel_transformer.py

```diff
@@ -0,0 +1,269 @@
+# coding=utf-8
+# Copyright 2020 The Trax Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Lint as: python3
+"""Funnel Transformer model.
+
+Funnel-Transformer: Filtering out Sequential Redundancy for Efficient
+Language Processing https://arxiv.org/abs/2006.03236 """
+from trax import layers as tl
+from trax.layers.assert_shape import assert_shape
+from trax.models.transformer import _EncoderBlock, _FeedForwardBlock
+
+
+def _InternalMaxPool(arr):
+  shape = arr.shape
+  arr = arr.reshape((*shape[:-1], int(shape[-1] / 2), 2))
+  arr = arr.max(axis=-1, keepdims=False)
+  return arr
+
+
+@assert_shape('bld->bSd')
+def PoolLayer(pool_layer=tl.AvgPool,
+              pool_size=(2,),
+              strides=(2,),
+              separate_cls=True):
+  return tl.Serial(
+      tl.Branch(
+          tl.Fn('select_cls_token', lambda x: x[:, :1, :]),
+          tl.Serial(
+              tl.Fn('rest_tokens', lambda x: x[:, 1:, :]),
+              pool_layer(pool_size, strides)
+          )
+      ),
+      tl.Concatenate(axis=1)
+  ) if separate_cls else pool_layer(pool_size, strides)
+
+
+def _Upsample(short, masks, long):
+  factor = -(-long.shape[1] // short.shape[1])  # ceil division
```

Comment:
I don't think this works as upsampling; short counterexample below. I think it would be reasonable just to assert that input length is always divisible by pool_size wherever we use pooling (input length is usually a power of two anyway). Usually the input length in training/evaluation is a power of 2 anyway, I think.

Counterexample: input_size=31, pool_size=2, single decoding block. AvgPool with default padding ("valid", not "same") will produce tensor of length 15. After this downsampling block we will have upsampling with factor = 3. This is out of sync with pool_size - we will get tokens mapped to wrong positions in "short.repeat".
While this counterexample would be fixed by changing the padding to "same", I have a feeling that we should just assert that input length at every stage will be divisible by pool_size. What do you think?

Reply:
Asserting that input length is divisible by pool_size would be fine for FunnelTransformer with upsampling, but we use the same pooler for the version without it. We could modify pooler to consider whether its output will be upsampled later, but I'm not sure if it's worth it.

Reply:
What could be done fairly easily is to insert in _Upsample function check that "if long.shape[1] % short.shape[1] != 0: raise ValueError('message')." Then you don't need to modify anything but _Upsample function; downsampling isn't touched at all, and we can be sure that upsampling works correctly.
The current issue I have with this _Upsample function is that it computes wrong results (see my counterexample), and I think that throwing an exception is much better than silently returning wrong results.

Adding this check is also much easier than correcting the implementation - this would involve passing a pool_size/stride to _Upsample to be used in place of 'factor', and some padding (see the counterexample in my previous comment). I think that correcting the implementation may not be worth the effort, but adding an assert is worth it.

(Also, while it isn't necessary, I would consider adding this kind of assert even during downsampling. Let's consider the case when we have only downsampling, with pool_size=2, and input length not divisible by 2. Current downsampling simply throws away the last token (due to a padding "valid", which is the default), which is kind of a strange behaviour.)